File: app/src/main/java/com/yashovardhan99/healersdiary/Activities/HealingLogs.java
package com.yashovardhan99.healersdiary.Activities;

import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.format.DateUtils;
import android.util.Log;
import android.view.MenuItem;
import android.widget.TextView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentChange;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QuerySnapshot;
import com.yashovardhan99.healersdiary.Adapters.HealingLogAdapter;
import com.yashovardhan99.healersdiary.Objects.Healing;
import com.yashovardhan99.healersdiary.R;

import java.text.DateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Objects;

import javax.annotation.Nullable;

public class HealingLogs extends AppCompatActivity {

    private RecyclerView.Adapter mAdapter;
    CollectionReference logs;
    //arraylist setup
    final ArrayList<Healing> healings = new ArrayList<>();
    FirebaseFirestore db;
    public int this_day, this_month, last_month;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_healing_logs);

        //toolbar setup
        android.support.v7.widget.Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setTitle(getIntent().getStringExtra("PATIENT_NAME"));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        //setting counters
        this_day = 0;
        this_month = 0;
        last_month = 0;

        //initialize firestore
        db = FirebaseFirestore.getInstance();
        logs = db.collection("users")
                .document(Objects.requireNonNull(FirebaseAuth.getInstance().getCurrentUser()).getUid())
                .collection("patients")
                .document(getIntent().getStringExtra(MainActivity.PATIENT_UID))
                .collection("healings");


        //collection reference to all healing logs for this patient
        logs.addSnapshotListener(new EventListener<QuerySnapshot>() {
            @Override
            public void onEvent(@Nullable QuerySnapshot queryDocumentSnapshots, @Nullable FirebaseFirestoreException e) {
                DateFormat df = DateFormat.getDateTimeInstance();
                if(e!=null){
                    Log.d("FIRESTORE",e.getLocalizedMessage());
                    return;
                }
                Log.d("FIRESTORE","Logs fetched");

                //fetching changes to this patient's healing collection
                if (queryDocumentSnapshots != null) {
                    for(DocumentChange dc : queryDocumentSnapshots.getDocumentChanges()){

                        Date date = dc.getDocument().getDate("Date");
                        Healing healing = new Healing();
                        healing.date = df.format(date);
                        healing.Uid = dc.getDocument().getId();

                        //get today's date
                        Calendar day =Calendar.getInstance();
                        day.setTime(new Date());
                        Calendar cal = Calendar.getInstance();
                        cal.setTime(date);

                        Boolean today = DateUtils.isToday(cal.getTimeInMillis());//check today
                        Boolean month = cal.get(Calendar.MONTH) == day.get(Calendar.MONTH);//check this month
                        //for last month
                        cal.add(Calendar.MONTH, 1);
                        Boolean last = cal.get(Calendar.MONTH) == day.get(Calendar.MONTH);

                        switch (dc.getType()){
                            case ADDED:
                                //new healing added
                                healings.add(0,healing);
                                mAdapter.notifyItemInserted(0);
                                //increment counters appropriately
                                if(today)
                                    this_day++;
                                if(month)
                                    this_month++;
                                if(last)
                                    last_month++;
                                break;
                            case REMOVED:
                                //healing removed
                                //decrement counters
                                if(today)
                                    this_day--;
                                if(month)
                                    this_month--;
                                if(last)
                                    last_month--;
                                int pos = healings.indexOf(healing);
                                if(pos<0)
                                    break;
                                healings.remove(healing);
                                mAdapter.notifyItemRemoved(pos);
                                break;
                        }
                        //NOTE - healing data cannot be modified
                        updateTextFields();
                    }
                }
            }
        });

        //recycler view setup
        RecyclerView mRecyclerView = findViewById(R.id.healingLogRecyclerView);
        mRecyclerView.setHasFixedSize(false);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mAdapter = new HealingLogAdapter(healings);
        mRecyclerView.setAdapter(mAdapter);

        //for the divider lines between each record
        DividerItemDecoration itemDecoration = new DividerItemDecoration(mRecyclerView.getContext(),DividerItemDecoration.VERTICAL);
        mRecyclerView.addItemDecoration(itemDecoration);

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case android.R.id.home:
                finish();
                return true;
                //this is to prevent empty activity from loading up on up button press which could lead to app crash
            default: return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        switch (item.getTitle().toString()){
            case "Delete":
                //record to be deleted
                deleteHealing(item.getGroupId());//deleted from firestore
                healings.remove(item.getGroupId());//removed from list
                Snackbar.make(findViewById(R.id.healingLogRecyclerView),"Record Deleted",Snackbar.LENGTH_LONG);
                mAdapter.notifyItemRemoved(item.getGroupId());//updated adapter
                return true;
        }
        return super.onContextItemSelected(item);
    }

    void deleteHealing(int id){
        //deletes healing from firestore
        Healing healing = healings.get(id);
        logs.document(healing.getUid())
                .delete();
        //now changing amount due
        Objects.requireNonNull(logs.getParent()).get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
            @Override
            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                double due = 0.00, rate = 0.00;
                if(task.isSuccessful() && task.getResult()!=null){
                    if(task.getResult().contains("Rate")){
                       rate = Objects.requireNonNull(task.getResult().getDouble("Rate"));
                    }
                    if(task.getResult().contains("Due"))
                        due = Objects.requireNonNull(task.getResult().getDouble("Due"));
                }
                else
                    Log.d("FIRESTORE", String.valueOf(task.getException()));

                logs.getParent().update("Due",due-rate);//updating amount due
            }
        });
    }
    void updateTextFields(){
        TextView today = findViewById(R.id.todayHealings);
        TextView month = findViewById(R.id.thisMonthHealings);
        TextView last = findViewById(R.id.lastMonthHealings);

        Resources res = getResources();

        today.setText(res.getQuantityString(R.plurals.healing,this_day,this_day,getString(R.string.today)));
        month.setText(res.getQuantityString(R.plurals.healing,this_month,this_month,getString(R.string.this_month)));
        last.setText(res.getQuantityString(R.plurals.healing, last_month, last_month, getString(R.string.last_month)));
    }
}

File: app/src/main/java/com/yashovardhan99/healersdiary/activities/HealingLogs.kt
package com.yashovardhan99.healersdiary.activities

import android.os.Bundle
import android.text.format.DateUtils
import android.util.Log
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentChange
import com.google.firebase.firestore.EventListener
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.yashovardhan99.healersdiary.R
import com.yashovardhan99.healersdiary.adapters.HealingLogAdapter
import com.yashovardhan99.healersdiary.objects.Healing
import java.text.DateFormat
import java.util.*

class HealingLogs : AppCompatActivity() {

    private lateinit var mAdapter: RecyclerView.Adapter<*>
    private lateinit var logs: CollectionReference
    //arraylist setup
    private val healings = ArrayList<Healing>()
    private lateinit var db: FirebaseFirestore
    private var thisDay: Int = 0
    private var thisMonth: Int = 0
    private var lastMonth: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_healing_logs)

        //toolbar setup
        val toolbar = findViewById<androidx.appcompat.widget.Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.title = intent.getStringExtra("PATIENT_NAME")
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        //setting counters
        thisDay = 0
        thisMonth = 0
        lastMonth = 0

        //initialize firestore
        db = FirebaseFirestore.getInstance()
        logs = db.collection(MainActivity.USERS)
                .document(FirebaseAuth.getInstance().currentUser!!.uid)
                .collection("patients")
                .document(intent.getStringExtra(MainActivity.PATIENT_UID))
                .collection("healings")


        //collection reference to all healing logs for this patient
        logs.limit(100)
                .orderBy("Date", Query.Direction.DESCENDING)
                .addSnapshotListener(EventListener { queryDocumentSnapshots, e ->
                    val df = DateFormat.getDateTimeInstance()
                    if (e != null) {
                        Log.d("FIRESTORE", e.localizedMessage)
                        return@EventListener
                    }
                    Log.d("FIRESTORE", "Logs fetched")

                    //fetching changes to this patient's healing collection
                    if (queryDocumentSnapshots != null) {
                        loop@ for (dc in queryDocumentSnapshots.documentChanges) {

                            val date = dc.document.getDate("Date")
                            val healing = Healing(df.format(date), dc.document.id)

                            //get today's date
                            val day = Calendar.getInstance()
                            day.time = Date()
                            val cal = Calendar.getInstance()
                            cal.time = date

                            val today = DateUtils.isToday(cal.timeInMillis)//check today
                            val month = cal.get(Calendar.MONTH) == day.get(Calendar.MONTH)//check this month
                            //for last month
                            cal.add(Calendar.MONTH, 1)
                            val last = cal.get(Calendar.MONTH) == day.get(Calendar.MONTH)

                            when (dc.type) {
                                DocumentChange.Type.ADDED -> {
                                    //new healing added
                                    healings.add(healing)
                                    mAdapter.notifyItemInserted(healings.size - 1)
                                    //increment counters appropriately
                                    if (today)
                                        thisDay++
                                    if (month)
                                        thisMonth++
                                    if (last)
                                        lastMonth++
                                }
                                DocumentChange.Type.REMOVED -> {
                                    //healing removed
                                    //decrement counters
                                    if (today)
                                        thisDay--
                                    if (month)
                                        thisMonth--
                                    if (last)
                                        lastMonth--
                                    val pos = healings.indexOf(healing)
                                    if (pos >= 0) {
                                        healings.remove(healing)
                                        mAdapter.notifyItemRemoved(pos)
                                    }
                                }
                                DocumentChange.Type.MODIFIED -> mAdapter.notifyItemChanged(healings.indexOf(healing))
                            }
                            //NOTE - healing data cannot be modified
                            updateTextFields()
                        }
                    }
                })

        //recycler view setup
        val mRecyclerView = findViewById<RecyclerView>(R.id.healingLogRecyclerView)
        mRecyclerView.setHasFixedSize(false)
        mRecyclerView.layoutManager = LinearLayoutManager(this)
        mAdapter = HealingLogAdapter(healings)
        mRecyclerView.adapter = mAdapter

        //for the divider lines between each record
        val itemDecoration = DividerItemDecoration(mRecyclerView.context, DividerItemDecoration.VERTICAL)
        mRecyclerView.addItemDecoration(itemDecoration)

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }
            //this is to prevent empty activity from loading up on up button press which could lead to app crash
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        when (item.title.toString()) {
            "Delete" -> {
                //record to be deleted
                deleteHealing(item.groupId)//deleted from firestore
                healings.removeAt(item.groupId)//removed from list
                Snackbar.make(findViewById(R.id.healingLogRecyclerView), R.string.deleted, Snackbar.LENGTH_LONG)
                mAdapter.notifyItemRemoved(item.groupId)//updated adapter
                return true
            }
        }
        return super.onContextItemSelected(item)
    }

    private fun deleteHealing(id: Int) {
        //deletes healing from firestore
        val (_, uid) = healings[id]
        logs.document(uid)
                .delete()
        //now changing amount due
        logs.parent!!.get().addOnCompleteListener { task ->
            var due = 0.00
            var rate = 0.00
            if (task.isSuccessful && task.result != null) {
                if (task.result!!.contains("Rate")) {
                    rate = task.result!!.getDouble("Rate")!!
                }
                if (task.result!!.contains("Due"))
                    due = task.result!!.getDouble("Due")!!
            } else
                Log.d("FIRESTORE", task.exception.toString())

            logs.parent!!.update("Due", due - rate)//updating amount due
        }
    }

    private fun updateTextFields() {
        val today = findViewById<TextView>(R.id.todayHealings)
        val month = findViewById<TextView>(R.id.thisMonthHealings)
        val last = findViewById<TextView>(R.id.lastMonthHealings)

        val res = resources

        today.text = res.getQuantityString(R.plurals.healing, thisDay, thisDay, getString(R.string.today))
        month.text = res.getQuantityString(R.plurals.healing, thisMonth, thisMonth, getString(R.string.this_month))
        last.text = res.getQuantityString(R.plurals.healing, lastMonth, lastMonth, getString(R.string.last_month))
    }
}